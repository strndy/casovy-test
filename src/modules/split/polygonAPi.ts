import { Split } from "../../types";

const POLYGON_API_KEY = "fpzp9R_q82pFzm1nwXWFNGlBpk6G6gOB";

export const fetchSplits = async (stockTicker: string) : Promise<Split[]> => {
    const url = `https://api.polygon.io/v3/reference/splits?`
    + `ticker=${stockTicker}` 
    + `&apiKey=${POLYGON_API_KEY}`;

    const response = await fetch(url);
    const data = await response.json();
    if (data.status != "OK") {
        throw new Error(`Failed to load splits for ${stockTicker}`);
    }
    data.results.forEach((s: Split) => s.execution_date = new Date(s.execution_date));
    return data.results as Split [];
}
