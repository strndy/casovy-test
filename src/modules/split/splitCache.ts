import fs from 'fs/promises';
import path from 'path';

export const cacheSplit = async (ticker: string, loadsplit: () => Promise<any>) => {
    const splitFilePath = path.join(__dirname, `../../../splitDb/splits_${ticker}.json`);
    try {
        
        const stats = await fs.stat(splitFilePath);
        const oneWeekAgo = new Date();
        oneWeekAgo.setDate(oneWeekAgo.getDate() - 7);
        
        if (stats.mtime < oneWeekAgo) {
            console.log(`Cache is older than a week for ${ticker}, reloading`);
            throw new Error('Cache is older than a week');
        }

        const cached = await fs.readFile(splitFilePath, 'utf8');
        const parsed = JSON.parse(cached);
        console.log(`Loaded ${parsed.results.length} splits for ${ticker} from cache`);
        return parsed.split;
    } catch (error) {
        const splits = await loadsplit();
        await fs.writeFile(splitFilePath, JSON.stringify(splits, null, 2));
        console.log(`Cached ${splits.results.length} splits for ${ticker}`);
        return splits;
    }
}