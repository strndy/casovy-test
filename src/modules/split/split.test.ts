import path from "path";
import fs from 'fs';

import { fetchSplits } from "./polygonAPi";
import { loadSplits, splitCache } from "./splitCache";

describe('split', () => {
    it('should cache splits', async () => {
        const splits = await splitCache("RLGT", () => fetchSplits("RLGT"));
        expect(splits).toBeDefined();
        expect(splits.length).toBe(1);
        expect(splits).toMatchSnapshot();

        try {   
            fs.unlinkSync(path.join(__dirname, `../../../splitDb/splits_RLGT.json`));
        } catch (error) {}

        const splits2 = await splitCache("RLGT", () => fetchSplits("RLGT"));
        expect(splits2).toBeDefined();
        expect(splits2.length).toBe(1);
        expect(splits2).toMatchSnapshot();

        const splits3 = await splitCache("RLGT", () => fetchSplits("RLGT"));
        expect(splits3).toBeDefined();
        expect(splits3.length).toBe(1);
        expect(JSON.stringify(splits)).toEqual(JSON.stringify(splits3));
        expect(splits3).not.toBe(splits); // Ensure they are not the same object reference
    });

    it('should load multiple stocks splits from polygon', async () => {
        const splits = await loadSplits(["TSLA", "AAPL"]);
        expect(splits).toBeDefined();
        expect(splits.length).toBe(2);
        expect(splits).toMatchSnapshot();
    });
});
