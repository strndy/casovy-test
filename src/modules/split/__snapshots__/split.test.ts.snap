// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`split should cache splits 1`] = `
[
  {
    "execution_date": 2005-10-21T00:00:00.000Z,
    "id": "E6fab3436e2dcdb6f42f2a26f173a8bf4a7576211011a736667701c307ef1ff86",
    "split_from": 1,
    "split_to": 3.5,
    "ticker": "RLGT",
  },
]
`;

exports[`split should cache splits 2`] = `
[
  {
    "execution_date": 2005-10-21T00:00:00.000Z,
    "id": "E6fab3436e2dcdb6f42f2a26f173a8bf4a7576211011a736667701c307ef1ff86",
    "split_from": 1,
    "split_to": 3.5,
    "ticker": "RLGT",
  },
]
`;

exports[`split should load multiple stocks splits from polygon 1`] = `
[
  [
    {
      "execution_date": 2022-08-25T00:00:00.000Z,
      "id": "Ea58e4c7785a1f9b68c3641137e7319bd7f7ff71d006d3132e0d0141b03860eb9",
      "split_from": 1,
      "split_to": 3,
      "ticker": "TSLA",
    },
    {
      "execution_date": 2020-08-31T00:00:00.000Z,
      "id": "Ec6bfc0213a18bfe35770ec1c95dc1f2ee90e99ab3dc94aad5e9adbe54ad2d6d2",
      "split_from": 1,
      "split_to": 5,
      "ticker": "TSLA",
    },
  ],
  [
    {
      "execution_date": 2020-08-31T00:00:00.000Z,
      "id": "E36416cce743c3964c5da63e1ef1626c0aece30fb47302eea5a49c0055c04e8d0",
      "split_from": 1,
      "split_to": 4,
      "ticker": "AAPL",
    },
    {
      "execution_date": 2014-06-09T00:00:00.000Z,
      "id": "E91a6b74ca1a9dcbce26a1f34e24ae26ba2c6359822ccf901ecd827f419137654",
      "split_from": 1,
      "split_to": 7,
      "ticker": "AAPL",
    },
    {
      "execution_date": 2005-02-28T00:00:00.000Z,
      "id": "E90a77bdf742661741ed7c8fc086415f0457c2816c45899d73aaa88bdc8ff6025",
      "split_from": 1,
      "split_to": 2,
      "ticker": "AAPL",
    },
  ],
]
`;
